model/routing/ResourceUrlBuilder.php

```diff
@@ -15,81 +15,87 @@
  *
- * Copyright (c) 2018 (original work) Open Assessment Technologies SA ;
+ * Copyright (c) 2018-2022 (original work) Open Assessment Technologies SA.
+ *
+ * @author Gyula Szucs <gyula@taotesting.com>
  */
 
+declare(strict_types=1);
+
 namespace oat\taoBackOffice\model\routing;
 
-use oat\generis\model\OntologyAwareTrait;
-use oat\oatbox\service\ConfigurableService;
+use LogicException;
+use oat\tao\model\menu\Tree;
+use InvalidArgumentException;
+use core_kernel_classes_Class;
+use oat\tao\model\menu\Section;
+use core_kernel_classes_Resource;
 use oat\tao\model\menu\MenuService;
 use oat\tao\model\menu\Perspective;
-use oat\tao\model\menu\Section;
+use oat\oatbox\service\ConfigurableService;
 
-/**
- * @author Gyula Szucs <gyula@taotesting.com>
- */
 class ResourceUrlBuilder extends ConfigurableService
 {
-    use OntologyAwareTrait;
-
-    const SERVICE_ID = 'taoBackOffice/resourceUrlBuilder';
+    public const SERVICE_ID = 'taoBackOffice/resourceUrlBuilder';
 
     /**
      * Builds a full URL for a resource
      *
-     * @param \core_kernel_classes_Resource $resource
      * @return string
      */
-    public function buildUrl(\core_kernel_classes_Resource $resource)
+    public function buildUrl(core_kernel_classes_Resource $resource)
     {
-        if (!$resource->exists()) {
-            throw new \InvalidArgumentException('The requested resource does not exist or has been deleted');
-        }
+        $resourceClass = $resource->getClass($resource);
 
-        /** @var Perspective $perspective */
-        /** @var Section $section */
+        if (!$resource->exists() && !$resourceClass->exists()) {
+            throw new InvalidArgumentException('The requested resource does not exist or has been deleted');
+        }
 
-        $resourceClass = $resource->isClass()
-            ? $this->getClass($resource)
-            : array_values($resource->getTypes())[0];
+        if (!$resource->isClass()) {
+            $resourceClass = array_values($resource->getTypes())[0];
+        }
 
+        /** @var Perspective $perspective */
         foreach (MenuService::getAllPerspectives() as $perspective) {
+            /** @var Section $section */
             foreach ($perspective->getChildren() as $section) {
                 if ($this->isSectionApplicable($resourceClass, $section)) {
                     return $this->getBackofficeUrl($perspective, $section, $resource);
                 }
             }
         }
 
-        throw new \LogicException('No url could be built for "' . $resource->getUri() . '"');
+        throw new LogicException(
+            sprintf(
+                'No url could be built for "%s"',
+                $resource->getUri()
+            )
+        );
     }
 
     /**
      * Generates the actual URL based on perspective and section
-     * @param Perspective $perspective
-     * @param Section $section
-     * @param \core_kernel_classes_Resource $resource
+     *
+     * @return string
      */
```

Comment:
suggestion: We are already using return type hints in this class, I think you can directly put a return value here instead of the PHPDoc annotation.
```suggestion
     */
```

Reply:
_(GitHub won't let me put the suggestion spawning the docblock and the method declaration itself for some reason :man_shrugging: :point_down: )_

Reply:
https://github.com/oat-sa/extension-tao-backoffice/pull/144#discussion_r872557560